pkcs11module.c

```diff
@@ -1322,30 +1471,350 @@ PHP_METHOD(Module, C_CreateObject) {
+
+            (void)h; /* unused */
+
+            if (key) /* ignore string key */
+                continue;
+
+            if (Z_TYPE_P(value) != IS_ARRAY)
+                continue;
+            /*
+             * typedef struct CK_ATTRIBUTE {
+             *   CK_ATTRIBUTE_TYPE type;
+             *   CK_VOID_PTR pValue;
+             *   CK_ULONG ulValueLen;
+             * } CK_ATTRIBUTE;
+             *
+             * value['type'](int), value['Value'](string)
+             */
+            ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(value), hi, keyi, valuei) {
+                (void)hi; /* unused */
+
+                if (keyi && (strcmp(ZSTR_VAL(keyi), "type") == 0)) {
+                    if (Z_TYPE_P(valuei) != IS_LONG)
+                        continue;
+                    assert(pTemplate[i].type == Z_LVAL_P(valuei));
+                    if ((pTemplate[i].ulValueLen == CK_UNAVAILABLE_INFORMATION) ||
+                        (pTemplate[i].ulValueLen < 1))
+                        continue;
+                    pTemplate[i].pValue = (CK_BYTE_PTR)ecalloc(1, pTemplate[i].ulValueLen);
+                }
+            } ZEND_HASH_FOREACH_END();
+            i++;
+        } ZEND_HASH_FOREACH_END();
+    }
+
+    /* fetch the content */
+    rv = objval->functionList->C_GetAttributeValue(sessionobjval->session,
+                                                   hObject, pTemplate, ulCount);
+    /*
+     * Note that the error codes CKR_ATTRIBUTE_SENSITIVE, CKR_ATTRIBUTE_TYPE_INVALID,
+     * and CKR_BUFFER_TOO_SMALL do not denote true errors for C_GetAttributeValue.
+     * If a call to C_GetAttributeValue returns any of these three values, then the
+     * call MUST nonetheless have processed every attribute in the template supplied
+     * to C_GetAttributeValue.
+     * Each attribute in the template whose value can be returned by the call to
+     * C_GetAttributeValue will be returned by the call to C_GetAttributeValue.
+     */
+    switch(rv) {
+        case CKR_ATTRIBUTE_SENSITIVE:
+        case CKR_ATTRIBUTE_TYPE_INVALID: /* one of the requested attributed does not exist */
+        case CKR_BUFFER_TOO_SMALL:
+        case CKR_OK:
+            break; /* ok */
+        default:
+            goto fini; /* free pTemplate, infos, pTemplate[x].pValue */
+            RETURN_LONG(rv); // placeholder, will be performed into the fini section
+    }
+
+    zval O;
+    array_init(&O);
+    for(CK_ULONG k = 0; k < ulCount; k++) {
+        zval zva;
+        if ((pTemplate[k].ulValueLen == CK_UNAVAILABLE_INFORMATION) ||
+            (pTemplate[k].ulValueLen < 1))
+            continue;
+        array_init(&zva);
+        add_assoc_long(&zva, "type", pTemplate[k].type);
+        add_assoc_stringl(&zva, "Value", pTemplate[k].pValue, pTemplate[k].ulValueLen);
+        if (infos[k])
+            add_assoc_string(&zva, "info", infos[k]);
+
+        add_next_index_zval(&O, &zva);
+    }
+    add_assoc_zval(template, "Object", &O);
```

Comment:
I'm not sure that this sticks to our desire to keep the original OASIS API for the C_* functions. Do you have a working example? I don't see a test for it.

Reply:
I'll check if I can apply the same loop than getAttributeValue() from the Object class (pkcs11object.c). If yes, I agree that we'll get a much simpler code.

Reply:
FYI, the running example is here:
   https://github.com/gamringer/php-pkcs11/pull/18/commits/295a2169580f226c1510f379f430908022626499#diff-bf9793d94afafb1194b6e68d92bdd968b9a3e10f2c6632f40022c4d2756be4deR27

Reply:
The OASIS gap with getAttributeValue() seems to be huge, I cannot get it closer with both logics. I feel we should keep it based on the current proposal.

Reply:
I'll post an update: last night I found a way to use parseTemplate() for this case too.

Reply:
A closer design to the OASIS code using parseTemplate() is available with this new serie:
   https://github.com/gamringer/php-pkcs11/pull/18/commits/b5366ee6884e5d97be584104721b07a35d00bab2

The example of usage is available here:
   https://github.com/gamringer/php-pkcs11/pull/18/commits/756c72677c75a4911e5714ca7cac261fc8918fd3